src/lib.rs

```diff
@@ -5,651 +5,741 @@
-        true
-    } else {
-        match c {
-            '!' |
-            '#' |
-            '$' |
-            '&' |
-            '-' |
-            '^' |
-            '.' |
-            '+' |
-            '_' => true,
-            _ => false
+macro_rules! mime_constant {
+    ($id:ident, $src:expr, $slash:expr) => (
+        mime_constant!($id, $src, $slash, None);
+    );
+    ($id:ident, $src:expr, $slash:expr, $plus:expr) => (
+        mime_constant!(FULL $id, $src, $slash, $plus, Params::None);
+    );
+
+    ($id:ident, $src:expr, $slash:expr, $plus:expr, $params:expr) => (
+        mime_constant!(FULL $id, $src, $slash, $plus, Params::Utf8($params));
+    );
+
+
+    (FULL $id:ident, $src:expr, $slash:expr, $plus:expr, $params:expr) => (
+        #[doc = $src]
+        pub const $id: Mime = Mime {
+            source: Source::Atom(__Atoms::$id as u8, $src),
+            slash: $slash,
+            plus: $plus,
+            params: $params,
+        };
+    )
+}
+
+
+#[cfg(test)]
+macro_rules! mime_constant_test {
+    ($id:ident, $src:expr, $slash:expr) => (
+        mime_constant_test!($id, $src, $slash, None);
+    );
+    ($id:ident, $src:expr, $slash:expr, $plus:expr) => (
+        mime_constant_test!(FULL $id, $src, $slash, $plus, Params::None);
+    );
+
+    ($id:ident, $src:expr, $slash:expr, $plus:expr, $params:expr) => (
+        mime_constant_test!(FULL $id, $src, $slash, $plus, Params::Utf8($params));
+    );
+
+    (FULL $id:ident, $src:expr, $slash:expr, $plus:expr, $params:expr) => ({
+        let __mime = $id;
+        let __slash = __mime.as_ref().as_bytes()[$slash];
+        assert_eq!(__slash, b'/', "{:?} has {:?} at slash position {:?}", __mime, __slash as char, $slash);
+        if let Some(plus) = __mime.plus {
+            let __c = __mime.as_ref().as_bytes()[plus];
+            assert_eq!(__c, b'+', "{:?} has {:?} at plus position {:?}", __mime, __c as char, plus);
+        } else {
+            assert!(!__mime.as_ref().as_bytes().contains(&b'+'), "{:?} forgot plus", __mime);
         }
-    }
+        if let Params::Utf8(semicolon) = __mime.params {
+            assert_eq!(__mime.as_ref().as_bytes()[semicolon], b';');
+            assert_eq!(&__mime.as_ref()[semicolon..], "; charset=utf-8");
+        } else if let Params::None = __mime.params {
+            assert!(!__mime.as_ref().as_bytes().contains(&b';'));
+        } else {
+            unreachable!();
+        }
+        __mime.atom()
+    })
+}
+
+
+mimes! {
+    STAR_STAR, "*/*", 1;
+
+    TEXT_PLAIN, "text/plain", 4;
+    TEXT_PLAIN_UTF_8, "text/plain; charset=utf-8", 4, None, 10;
+    TEXT_HTML, "text/html", 4;
+    TEXT_CSS, "text/css", 4;
+    TEXT_JAVSCRIPT, "text/javascript", 4;
+    TEXT_XML, "text/xml", 4;
+    TEXT_EVENT_STREAM, "text/event-stream", 4;
+
+    IMAGE_JPEG, "image/jpeg", 5;
+    IMAGE_GIF, "image/gif", 5;
+    IMAGE_PNG, "image/png", 5;
+    IMAGE_BMP, "image/bmp", 5;
+
+    APPLICATION_JSON, "application/json", 11;
```

Comment:
Oh, thanks for the link, that's news :). I definitely remember having to add the charset to make some implementations work, but I guess the implementation was technically wrong in that case.

So it's probably better to not add it to guide people to the correct one. When someone really needs it they can just write it out themselves.